src/script/interpreter.cpp

```diff
@@ -1279,18 +1279,28 @@ uint256 GetOutputsHash(const T& txTo)
 } // namespace
 
 template <class T>
-PrecomputedTransactionData::PrecomputedTransactionData(const T& txTo)
+void PrecomputedTransactionData::Init(const T& txTo)
 {
     // Cache is calculated only for transactions with witness
     if (txTo.HasWitness()) {
-        hashPrevouts = GetPrevoutHash(txTo);
-        hashSequence = GetSequenceHash(txTo);
-        hashOutputs = GetOutputsHash(txTo);
-        ready = true;
+        if (!ready) {
```

Comment:
nit, early return in L1284 `if (ready) return;`?

Reply:
The later commits in 17977 can't return early, because we may need to cache the hash of the amounts spent in this transaction (see https://github.com/bitcoin/bitcoin/pull/17977/commits/607843e0e1e8e09f11192e6489ecb22b6acdae79#diff-be2905e2f5218ecdbe4e55637dac75f3R1308). However, I think we can return early for non-witness transactions, so I've implemented that.

Reply:
For now that's ok, but https://github.com/bitcoin/bitcoin/pull/17977/commits/dd0622b8e2f8a324fdf73eb476f136ee556f3e68 relies on storing the spent UTXOs in txdata, even for non-witness outputs (it makes validation use the scriptPubKeys from there, to avoid duplicating them).

Reply:
I think the function can be re-ordered to first store the spent UTXOs for all transactions, and then do the witness-specific logic later (and exit early for non-witness txs).

Reply:
Sure; just clarifying why it wasn't done in the original commit. This is a triviality that can always be changed.